src/util.ts

```diff
@@ -480,63 +515,49 @@ export async function fetchCalldataDecoder (_data: Uint8Array | string, to: stri
     const supportedChain = cachedNetwork
       ? cachedNetwork
       : await fetchExternalNetworkForChainId(chainId);
-    if (supportedChain) {
-      const url = buildUrlForSupportedChainAndAddress({ supportedChain, address: to })
-      const response = await superagent
-        .get(url)
-        .catch(err => {
-          console.warn(`Fetching calldata from ${supportedChain.name} failed\n`, err)
-        })
-      if (
-        response &&
-        response.body &&
-        response.body.result
-      ) {
-        const abi = JSON.parse(response.body.result)
-        return {
-          def: Calldata.EVM.parsers.parseSolidityJSONABI(selector, abi),
-          abi
-        }
+
+    try {
+      if (supportedChain) {
+        const url = buildUrlForSupportedChainAndAddress({ supportedChain, address: to })
+        return await superagent
+          .get(url)
+          .then(res => {
+            if (res && res.body && res.body.result) {
+              const abi = JSON.parse(res.body.result)
+              const def = Calldata.EVM.parsers.parseSolidityJSONABI(selector, abi)
+              return { abi, def }
+            } else {
+              throw new Error('Server response was malformed')
+            }
+          }).catch(() => {
+            throw new Error('Fetching data from external network failed')
+          })
+      } else {
+        throw new Error(`Chain (id: ${chainId}) is not supported`)
       }
-      console.warn(
-        `Could not find selector in calldata for network: ${supportedChain.name}\n`,
-        'Falling back to 4byte\n',
-      );
+    } catch (err) {
+      console.warn(err.message, '\n', 'Falling back to 4byte');
     }
 
     // Fallback to checking 4byte
     const url = `https://www.4byte.directory/api/v1/signatures?hex_signature=0x${selector}`
-    const response = await superagent
+    return await superagent
       .get(url)
-      .catch(err => {
-        console.warn('Failed to fetch calldata from 4byte \n', err)
-      });
-    if (
-      response &&
-      response.body &&
-      response.body.results &&
-      response.body.results.length
-    ) {
-      const fourByteResults = response.body.results;
-      if (fourByteResults.length > 1) {
-        console.warn('WARNING: There are multiple results. Using the first one.');
-      }
-      const canonicalName = fourByteResults.sort((a, b) => {
-        const aTime = new Date(a.created_at).getTime();
-        const bTime = new Date(b.created_at).getTime();
-        return aTime - bTime;
-      })[0]?.text_signature;
-      return {
-        def: Calldata.EVM.parsers.parseCanonicalName(selector, canonicalName),
-        abi: fourByteResults
-      }
-    }
-    throw new Error('Could not find selector in calldata from 4byte')
-  }
-  catch (err) {
+      .then(res => {
+        if (res && res.body && res.body.results && res.body.results.length) {
+          const abi = res.body.results
+          const def = selectDefFrom4byteABI(abi, selector)
```

Comment:
I'm a bit confused by the return codes. If you get a null `def` from this function, you will be returning the ABI but no def? In this case we don't want to add calldata to the fw requests because it won't decode anything.

Also, is L560 even reachable now?

Reply:
I updated `selectDefFrom4byteABI()` to throw if it does not find `def` [259a926](https://github.com/GridPlus/gridplus-sdk/pull/433/commits/259a92633a15f6fa5676835026c2c171e6336130)

This should result in only returning `{abi,def}` if `def` is defined. 

L560 is reached if **both** requests fail for one of the many reasons we cover with all these error checks.